presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddExchangesBelowPartialAggregationOverGroupIdRuleSet.java

```diff
@@ -0,0 +1,349 @@
+    {
+        @Override
+        public Pattern<ExchangeNode> getPattern()
+        {
+            return WITHOUT_PROJECTION;
+        }
+
+        @Override
+        public Result apply(ExchangeNode exchange, Captures captures, Context context)
+        {
+            AggregationNode aggregation = captures.get(AGGREGATION);
+            GroupIdNode groupId = captures.get(GROUP_ID);
+            return transform(aggregation, groupId, context)
+                    .map(newAggregation -> {
+                        PlanNode newExchange = exchange.replaceChildren(ImmutableList.of(newAggregation));
+                        return Result.ofPlanNode(newExchange);
+                    })
+                    .orElseGet(Result::empty);
+        }
+    }
+
+    private abstract class BaseAddExchangesBelowExchangePartialAggregationGroupId
+            implements Rule<ExchangeNode>
+    {
+        @Override
+        public boolean isEnabled(Session session)
+        {
+            if (!isEnableStatsCalculator(session)) {
+                // Old stats calculator is not trust-worthy
+                return false;
+            }
+
+            return isEnableForcedExchangeBelowGroupId(session);
+        }
+
+        protected Optional<PlanNode> transform(AggregationNode aggregation, GroupIdNode groupId, Context context)
+        {
+            if (groupId.getGroupingSets().size() < 2) {
+                return Optional.empty();
+            }
+
+            Set<Symbol> groupingKeys = aggregation.getGroupingKeys().stream()
+                    .filter(symbol -> !groupId.getGroupIdSymbol().equals(symbol))
+                    .collect(toImmutableSet());
+
+            Multiset<Symbol> groupingSetHistogram = groupId.getGroupingSets().stream()
+                    .flatMap(Collection::stream)
+                    .collect(toImmutableMultiset());
+
+            if (!Objects.equals(groupingSetHistogram.elementSet(), groupingKeys)) {
+                // TODO handle the case when some aggregation keys are pass-through in GroupId (e.g. common in all grouping sets). However, this is never the case for ROLLUP.
+                // TODO handle the case when some grouping set symbols are not used in aggregation (possible?)
+                return Optional.empty();
+            }
+
+            double aggregationMemoryRequirements = estimateAggregationMemoryRequirements(groupingKeys, groupId, groupingSetHistogram, context);
+            if (isNaN(aggregationMemoryRequirements) || aggregationMemoryRequirements < maxPartialAggregationMemoryUsage.toBytes()) {
+                // Aggregation will be effective even without exchanges.
+                return Optional.empty();
+            }
+
+            List<Symbol> desiredHashSymbols = groupingSetHistogram.entrySet().stream()
+                    // Take only frequently used symbols
+                    .filter(entry -> entry.getCount() >= groupId.getGroupingSets().size() * GROUPING_SETS_SYMBOL_REQUIRED_FREQUENCY)
```

Comment:
what if all symbols are less frequent, shouldn't we return Optional.empty() here (short circuit)?